### PR TITLE
Add tests to check ha of controller

### PR DESCRIPTION
**Pull request by cbickel**
Add a test, that checks, that the hot standby controller is used as expected.

PG3#1031 started

### CODE REVIEW

**Comment by cbickel** [comment]
PG3#1031 is 🔵

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    with Matchers
+    with WaitFor
+    with WskTestHelpers
+    with ScalaFutures
+    with WskActorSystem {
+
+    implicit val wskprops = WskProps()
+    val wsk = new Wsk
+    val defaultAction = Some(TestUtils.getTestActionFilename("hello.js"))
+    val python = WhiskProperties.python
```
Comment (markusthoemmes) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: Unused.

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                // Produce some load on the system for 90 seconds. Kill the controller after 4 requests
+                val requestsBeforeRestart = doRequests(4, actionName)
+
+                // Kill the controller
+                val dockerHost = WhiskProperties.getBaseControllerHost() + ":" + WhiskProperties.getProperty(WhiskConfig.dockerPort)
+                restartComponent(dockerHost, "controller0")
+                // Wait until down
+                retry({
+                    isControllerAlive(0) shouldBe false
+                }, 10, Some(100.milliseconds))
```
Comment (markusthoemmes) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: Is 1 second of retrying really enough? I think we can safely bump this to 10 seconds or even a minute to rule out deployment related details to interfere with the test, wdyt?

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  def isControllerAlive(instance: Int): Boolean = {
+    require(instance >= 0 && instance < 2, "Controller instance not known.")
+
+    val host = WhiskProperties.getProperty("controller.hosts").split(",")(instance)
+    val port = WhiskProperties.getControllerBasePort + instance
+
+    val res = ping(host, port)
+    res.isDefined && res.get._1 == StatusCodes.OK && res.get._2 == "pong"
```
Comment (markusthoemmes) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: Is this the same as:

```scala
res == Some(StatusCodes.OK, "pong")
```

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+      assetHelper.withCleaner(wsk.action, actionName) { (action, _) =>
+        action.create(actionName, defaultAction)
+      }
+
+      // Produce some load on the system for 90 seconds. Kill the controller after 4 requests
+      val requestsBeforeRestart = doRequests(4, actionName)
+
+      // Kill the controller
+      val dockerHost = WhiskProperties.getBaseControllerHost() + ":" + WhiskProperties.getProperty(
```
Comment (markusthoemmes) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: Should this be defined up top?

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        WhiskConfig.dockerPort)
+      restartComponent(dockerHost, "controller0")
+      // Wait until down
+      retry({
+        isControllerAlive(0) shouldBe false
+      }, 100, Some(100.milliseconds))
+      // Check that second controller is still up
+      isControllerAlive(1) shouldBe true
+
+      val requestsAfterRestart = doRequests(96, actionName)
```
Comment (markusthoemmes) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: Would it make sense to formulate this as `100 - 4` (define a `val totalRequests = 100` above?)

**Pull request merged**